File: split/checkoutervalidity.pas
procedure checkoutervalidity;
var
  p: halfword;
  q: halfword;
begin
  if scannerstatus <> 0 then
  begin
    deletionsallowed := false;
    {337:}
if curcs <> 0 then
    begin
      if (curinput.statefield = 0) or (curinput.namefield < 1) or (curinput.namefield > 17) then
      begin
        p := getavail;
        mem[p].hh.lh := 4095 + curcs;
        begintokenlist(p, 3);
      end;
      curcmd := 10;
      curchr := 32;
    end{:337}
;
    if scannerstatus > 1 then{338:}

    begin
      runaway;
      if curcs = 0 then
      begin
        if interaction = 3 then;
        printnl(262);
        print(604);
      end
      else
      begin
        curcs := 0;
        begin
          if interaction = 3 then;
          printnl(262);
          print(605);
        end;
      end;
      print(606);
      {339:}
p := getavail;
      case scannerstatus of
        2:
        begin
          print(570);
          mem[p].hh.lh := 637;
        end;
        3:
        begin
          print(612);
          mem[p].hh.lh := partoken;
          longstate := 113;
        end;
        4:
        begin
          print(572);
          mem[p].hh.lh := 637;
          q := p;
          p := getavail;
          mem[p].hh.rh := q;
          mem[p].hh.lh := 6710;
          alignstate := -1000000;
        end;
        5:
        begin
          print(573);
          mem[p].hh.lh := 637;
        end;
      end;
      begintokenlist(p, 4){:339}
;
      print(607);
      sprintcs(warningindex);
      begin
        helpptr := 4;
        helpline[3] := 608;
        helpline[2] := 609;
        helpline[1] := 610;
        helpline[0] := 611;
      end;
      error;
    end{:338}

    else
    begin
      begin
        if interaction = 3 then;
        printnl(262);
        print(598);
      end;
      printcmdchr(105, curif);
      print(599);
      printint(skipline);
      begin
        helpptr := 3;
        helpline[2] := 600;
        helpline[1] := 601;
        helpline[0] := 602;
      end;
      if curcs <> 0 then
        curcs := 0
      else
        helpline[2] := 603;
      curtok := 6713;
      inserror;
    end;
    deletionsallowed := true;
  end;
end;

File: cpp/checkoutervalidity.cpp
#include "checkoutervalidity.h"
#include "print.h"
#include "printcmdchr.h"
#include "printcs.h"
#include "printint.h"
#include "inserror.h"
#include "printnl.h"
#include "getavail.h"
#include "begintokenlist.h"
#include "sprintcs.h"
#include "runaway.h"
#include "error.h"
#include "texte.h"

void checkoutervalidity(void)
{
	if (scannerstatus)
	{
		deletionsallowed = false;
		if (curcs)
		{
			if (state == token_list || txt(name) < 1 || txt(name) > 17)
			{
				auto p = getavail();
				info(p) = cs_token_flag+curcs;
				back_list(p);
			}
			curcmd = spacer;
			curchr = ' ';
		}
		if (scannerstatus > skipping)
		{
			runaway();
			if (curcs == 0)
			{
				printnl("! ");
				print("File ended");
			}
			else
			{
				curcs = 0;
				printnl("! ");
				print("Forbidden control sequence found");
			}
			print(" while scanning ");
			auto p = getavail();
			decltype(p) q;
			switch (scannerstatus)
			{
				case defining:
					print("definition");
					info(p) = right_brace_token+'}';
					break;
				case matching:
					print("use");
					info(p) = partoken;
					longstate = outer_call;
					break;
				case aligning:
					print("preamble");
					info(p) = right_brace_token+'}';
					q = p;
					p = getavail();
					link(p) = q;
					info(p) = frozen_cr+cs_token_flag;
					alignstate = -1000000;
					break;
				case absorbing:
					print("text");
					info(p) = right_brace_token+'}';
			}
			ins_list(p);
			print(" of ");
			sprintcs(warningindex);
			{
				helpptr = 4;
				helpline[3] = "I suspect you have forgotten a `}', causing me";
				helpline[2] = "to read past where you wanted me to stop.";
				helpline[1] = "I'll try to recover; but if the error is serious,";
				helpline[0] = "you'd better type `E' or `X' now and fix your file.";
			}
			error();
		}
		else
		{
			printnl("! ");
			print("Incomplete ");
			printcmdchr(if_test, curif);
			print("; all text was ignored after line ");
			printint(skipline);
			helpptr = 3;
			helpline[2] = "A forbidden control sequence occurred in skipped text.";
			helpline[1] = "This kind of error happens when you say `\\if...' and forget";
			helpline[0] = "the matching `\\fi'. I've inserted a `\\fi'; this might work.";
			if (curcs)
				curcs = 0;
			else
				helpline[2] = "The file ended while I was skipping conditional text.";
			curtok = frozen_fi+cs_token_flag;
			inserror();
		}
		deletionsallowed = true;
	}
}
